File: app/src/main/java/pl/piotrskiba/angularowo/adapters/PlayerListAdapter.java
package pl.piotrskiba.angularowo.adapters;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import pl.piotrskiba.angularowo.IntegerVersionSignature;
import pl.piotrskiba.angularowo.R;
import pl.piotrskiba.angularowo.interfaces.PlayerClickListener;
import pl.piotrskiba.angularowo.models.Player;
import pl.piotrskiba.angularowo.models.PlayerList;
import pl.piotrskiba.angularowo.models.Rank;
import pl.piotrskiba.angularowo.utils.ColorUtils;
import pl.piotrskiba.angularowo.utils.GlideUtils;
import pl.piotrskiba.angularowo.utils.RankUtils;
import pl.piotrskiba.angularowo.utils.UrlUtils;

public class PlayerListAdapter extends RecyclerView.Adapter<PlayerListAdapter.PlayerViewHolder> {

    private final Context context;
    private final PlayerClickListener mClickListener;

    private PlayerList playerList;

    public PlayerListAdapter(Context context, PlayerClickListener clickListener){
        this.context = context;
        this.mClickListener = clickListener;
    }

    @NonNull
    @Override
    public PlayerViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        Context context = parent.getContext();
        LayoutInflater inflater = LayoutInflater.from(context);

        View view = inflater.inflate(R.layout.player_list_item, parent, false);

        return new PlayerViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull PlayerViewHolder holder, int position) {
        Player player = playerList.getPlayers().get(position);

        if(player.getUuid() != null) {
            Glide.with(context)
                    .load(UrlUtils.buildAvatarUrl(player.getUuid(), true))
                    .signature(new IntegerVersionSignature(GlideUtils.getSignatureVersionNumber(5)))
                    .placeholder(R.drawable.default_avatar)
                    .into(holder.mPlayerAvatar);
        }
        else{
            holder.mPlayerAvatar.setImageDrawable(context.getResources().getDrawable(R.drawable.default_avatar));
        }

        holder.mPlayerName.setText(player.getUsername());
        holder.mPlayerRank.setText(player.getRank().getName());

        int color = ColorUtils.getColorFromCode(player.getRank().getColorCode());
        ((ConstraintLayout) holder.mPlayerAvatar.getParent()).setBackgroundColor(color);

        if(player.isVanished())
            holder.mPlayerVanishIcon.setVisibility(View.VISIBLE);
        else
            holder.mPlayerVanishIcon.setVisibility(View.INVISIBLE);
    }

    @Override
    public int getItemCount() {
        if(playerList != null)
            return playerList.getPlayers().size();
        else
            return 0;
    }

    class PlayerViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        @BindView(R.id.iv_player_avatar)
        ImageView mPlayerAvatar;

        @BindView(R.id.tv_player_name)
        TextView mPlayerName;

        @BindView(R.id.tv_player_rank)
        TextView mPlayerRank;

        @BindView(R.id.iv_vanish_status)
        ImageView mPlayerVanishIcon;

        PlayerViewHolder(View itemView) {
            super(itemView);

            ButterKnife.bind(this, itemView);

            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(View view) {
            int pos = getAdapterPosition();
            mClickListener.onPlayerClick(view, playerList.getPlayers().get(pos));
        }
    }

    public void setPlayerList(PlayerList playerList){
        List<Player> sorted = new ArrayList<>();

        if(playerList.getPlayers() != null && !playerList.getPlayers().isEmpty()) {
            List<Rank> ranks = RankUtils.getAllRanks();

            for (Rank rank : ranks) {
                for (Player player : playerList.getPlayers()) {
                    if (player.getRank() != null && player.getRank().getName().equals(rank.getName()))
                        sorted.add(player);
                }
            }
        }

        this.playerList = new PlayerList(sorted);
        notifyDataSetChanged();
    }
}

File: app/src/main/java/pl/piotrskiba/angularowo/adapters/PlayerListAdapter.kt
package pl.piotrskiba.angularowo.adapters

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import butterknife.BindView
import butterknife.ButterKnife
import com.bumptech.glide.Glide
import pl.piotrskiba.angularowo.AppViewModel
import pl.piotrskiba.angularowo.IntegerVersionSignature
import pl.piotrskiba.angularowo.R
import pl.piotrskiba.angularowo.adapters.PlayerListAdapter.PlayerViewHolder
import pl.piotrskiba.angularowo.adapters.diffcallbacks.PlayerListDiffCallback
import pl.piotrskiba.angularowo.database.entity.Friend
import pl.piotrskiba.angularowo.interfaces.PlayerClickListener
import pl.piotrskiba.angularowo.models.Player
import pl.piotrskiba.angularowo.models.PlayerList
import pl.piotrskiba.angularowo.utils.ColorUtils.getColorFromCode
import pl.piotrskiba.angularowo.utils.GlideUtils.getSignatureVersionNumber
import pl.piotrskiba.angularowo.utils.RankUtils
import pl.piotrskiba.angularowo.utils.UrlUtils.buildAvatarUrl

class PlayerListAdapter(
        private val context: Context,
        private val mClickListener: PlayerClickListener,
        private val mViewModel: AppViewModel
) : RecyclerView.Adapter<PlayerViewHolder>() {

    private var playerList: PlayerList = PlayerList(ArrayList())

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PlayerViewHolder {
        val context = parent.context

        val inflater = LayoutInflater.from(context)
        val view = inflater.inflate(R.layout.player_list_item, parent, false)

        return PlayerViewHolder(view)
    }

    override fun onBindViewHolder(holder: PlayerViewHolder, position: Int) {
        val player = playerList.players[position]

        Glide.with(context)
                .load(buildAvatarUrl(player.skinUuid, true))
                .signature(IntegerVersionSignature(getSignatureVersionNumber(5)))
                .placeholder(R.drawable.default_avatar)
                .into(holder.mPlayerAvatar)

        holder.mPlayerName.text = player.username
        holder.mPlayerRank.text = player.rank.name

        val color = getColorFromCode(context, player.rank.colorCode)
        (holder.mPlayerAvatar.parent as ConstraintLayout).setBackgroundColor(color)

        if (player.isVanished)
            holder.mPlayerVanishIcon.visibility = View.VISIBLE
        else
            holder.mPlayerVanishIcon.visibility = View.INVISIBLE

        val friends = mViewModel.allFriends.value
        if (friends != null && friends.contains(Friend(player.uuid))) {
            holder.mPlayerHeartIcon.visibility = View.VISIBLE
        }
        else {
            holder.mPlayerHeartIcon.visibility = View.GONE
        }
    }

    override fun getItemCount(): Int {
        return playerList.players.size
    }

    inner class PlayerViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {

        @BindView(R.id.iv_player_avatar)
        lateinit var mPlayerAvatar: ImageView

        @BindView(R.id.tv_player_name)
        lateinit var mPlayerName: TextView

        @BindView(R.id.tv_player_rank)
        lateinit var mPlayerRank: TextView

        @BindView(R.id.iv_vanish_status)
        lateinit var mPlayerVanishIcon: ImageView

        @BindView(R.id.iv_heart)
        lateinit var mPlayerHeartIcon: ImageView

        override fun onClick(view: View) {
            mClickListener.onPlayerClick(view, playerList.players[bindingAdapterPosition])
        }

        init {
            ButterKnife.bind(this, itemView)
            itemView.setOnClickListener(this)
        }
    }

    fun setPlayerList(playerList: PlayerList) {
        val newPlayerList: MutableList<Player> = ArrayList()
        for (rank in RankUtils.allRanks) {
            for (player in playerList.players) {
                if (player.rank.name == rank.name) newPlayerList.add(player)
            }
        }

        val diffCallback = PlayerListDiffCallback(this.playerList, PlayerList(newPlayerList))
        val diffResult = DiffUtil.calculateDiff(diffCallback)

        this.playerList = PlayerList(newPlayerList)
        diffResult.dispatchUpdatesTo(this)
    }

}